### ISSUE TITLE
`NVRTCError` raised on Maxwell or earlier GPUs

### ISSUE THREAD

**Issue by samrere** [cupy/cupy#6615]
### Description

I've installed CuPy version '10.3.0' on Windows.
I created an array and want to convert to other datatype using array.astype(...) but it leads to error

### To Reproduce

**Code**
```py
import cupy as cp
a=cp.array([1, 2]).astype(cp.float32)
```

**Error message**
```py
---------------------------------------------------------------------------
NVRTCError                                Traceback (most recent call last)
D:\anaconda3\lib\site-packages\cupy\cuda\compiler.py in compile(self, options, log_stream)
    684                     nvrtc.addNameExpression(self.ptr, ker)
--> 685             nvrtc.compileProgram(self.ptr, options)
    686             mapping = None

cupy_backends\cuda\libs\nvrtc.pyx in cupy_backends.cuda.libs.nvrtc.compileProgram()

cupy_backends\cuda\libs\nvrtc.pyx in cupy_backends.cuda.libs.nvrtc.compileProgram()

cupy_backends\cuda\libs\nvrtc.pyx in cupy_backends.cuda.libs.nvrtc.check_status()

NVRTCError: NVRTC_ERROR_COMPILATION (6)

During handling of the above exception, another exception occurred:

CompileException                          Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_129232/2912708919.py in <module>
      1 import cupy as cp
----> 2 a=cp.array([1, 2]).astype(cp.float32)

cupy\_core\core.pyx in cupy._core.core.ndarray.astype()

cupy\_core\core.pyx in cupy._core.core.ndarray.astype()

cupy\_core\_kernel.pyx in cupy._core._kernel.ufunc.__call__()

cupy\_core\_kernel.pyx in cupy._core._kernel.ufunc._get_ufunc_kernel()

cupy\_core\_kernel.pyx in cupy._core._kernel._get_ufunc_kernel()

cupy\_core\_kernel.pyx in cupy._core._kernel._get_simple_elementwise_kernel()

cupy\_core\core.pyx in cupy._core.core.compile_with_cache()

D:\anaconda3\lib\site-packages\cupy\cuda\compiler.py in _compile_module_with_cache(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, jitify)
    490             name_expressions, log_stream, cache_in_memory)
    491     else:
--> 492         return _compile_with_cache_cuda(
    493             source, options, arch, cache_dir, extra_source, backend,
    494             enable_cooperative_groups, name_expressions, log_stream,

D:\anaconda3\lib\site-packages\cupy\cuda\compiler.py in _compile_with_cache_cuda(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, cache_in_memory, jitify)
    568     if backend == 'nvrtc':
    569         cu_name = '' if cache_in_memory else name + '.cu'
--> 570         ptx, mapping = compile_using_nvrtc(
    571             source, options, arch, cu_name, name_expressions,
    572             log_stream, cache_in_memory, jitify)

D:\anaconda3\lib\site-packages\cupy\cuda\compiler.py in compile_using_nvrtc(source, options, arch, filename, name_expressions, log_stream, cache_in_memory, jitify)
    319                 cu_file.write(source)
    320 
--> 321             return _compile(source, options, cu_path,
    322                             name_expressions, log_stream, jitify)
    323     else:

D:\anaconda3\lib\site-packages\cupy\cuda\compiler.py in _compile(source, options, cu_path, name_expressions, log_stream, jitify)
    303 
    304         try:
--> 305             compiled_obj, mapping = prog.compile(options, log_stream)
    306         except CompileException as e:
    307             dump = _get_bool_env_variable(

D:\anaconda3\lib\site-packages\cupy\cuda\compiler.py in compile(self, options, log_stream)
    700         except nvrtc.NVRTCError:
    701             log = nvrtc.getProgramLog(self.ptr)
--> 702             raise CompileException(log, self.src, self.name, options,
    703                                    'nvrtc' if not runtime.is_hip else 'hiprtc')
    704 

CompileException: D:\anaconda3\lib\site-packages\cupy\_core\include\cupy/carray.cuh(130): error: more than one conversion function from "half" to a built-in type applies:
            function "__half::operator float() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(204): here
            function "__half::operator short() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(222): here
            function "__half::operator unsigned short() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(225): here
            function "__half::operator int() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(228): here
            function "__half::operator unsigned int() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(231): here
            function "__half::operator long long() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(234): here
            function "__half::operator unsigned long long() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(237): here
            function "__half::operator __nv_bool() const"
D:\anaconda3\lib\site-packages\cupy\_core\include\cupy\_cuda\cuda-11\cuda_fp16.hpp(241): here

1 error detected in the compilation of "C:\Users\Sam\AppData\Local\Temp\tmp561lep1m\328eb3900d18cfb25d20bc9256e53d94036656ce.cubin.cu".
```

### Installation

Wheel (`pip install cupy-***`)

### Environment

```
OS                           : Windows-10-10.0.19044-SP0
Python Version               : 3.9.7
CuPy Version                 : 10.3.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.20.3
SciPy Version                : 1.7.1
Cython Build Version         : 0.29.28
Cython Runtime Version       : 0.29.24
CUDA Root                    : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.5
nvcc PATH                    : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.5\bin\nvcc.EXE
CUDA Build Version           : 11050
CUDA Driver Version          : 11050
CUDA Runtime Version         : 11050
cuBLAS Version               : (available)
cuFFT Version                : 10600
cuRAND Version               : 10207
cuSOLVER Version             : (11, 3, 2)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 5)
Thrust Version               : 101301
CUB Build Version            : 101301
Jitify Build Version         : 4a37de0
cuDNN Build Version          : (not loaded; try `import cupy.cuda.cudnn` first)
cuDNN Version                : (not loaded; try `import cupy.cuda.cudnn` first)
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : NVIDIA GeForce GTX 965M
Device 0 Compute Capability  : 52
Device 0 PCI Bus ID          : 0000:01:00.0
```

**Comment by samrere** [comment]
The following code with sparse matrix also leads to similar error:
```py
import cupy as cp
import cupyx.scipy as cpx

indptr = cp.array([0, 2, 3, 6])
indices = cp.array([0, 2, 2, 0, 1, 2])
data = cp.array([1.2, 2.2, 3.2, 4.2, 5.2, 6.2])
y = cpx.sparse.csr_matrix((data, indices, indptr))
```

**Comment by kmaehashi** [comment]
This issue seems a regression in v11.0.0b1/v10.3.0 release. The cause was here:
https://github.com/cupy/cupy/pull/6599/files#diff-9bb7de5d0f4d7dc8c55b1211f5021a1c2283259b0d0933133ac0a368f5c9812aR129-R132

```cuda
  __device__ float16 operator-() {
    return float16(-data_);
  }
```

**Comment by samrere** [comment]
Hi kmaehashi thanks for your reply! when I comment it out everything works. May I ask if it's safe to do so?

**Comment by kmaehashi** [comment]
This fix is specific to ROCm 5.0 so it should be safe to comment out as long as you are using CUDA.

**Comment by kmaehashi** [comment]
We are preparing a hot-fix release v10.3.1 to address this issue.
(no hot-fix planned for v11 pre-releases)

**Comment by kmaehashi** [comment]
We've published v10.3.1 hot-fix release: https://github.com/cupy/cupy/releases/tag/v10.3.1
@samrere Thanks again for reporting this!

**Comment by samrere** [comment]
Brilliant! Keep up the good work!
